code/datums/ai/sanity/_sanityloss_controller.dm

```diff
@@ -521,62 +599,151 @@
-		if(L.stat == DEAD)
-			continue
-		if(living_pawn.see_invisible < L.invisibility)
-			continue
-		if(!isturf(L.loc) && !ismecha(L.loc))
+	for(var/L in potential_enemies)
+		if(!CanTarget(L))
 			continue
-		weighted_list += L
-	for(var/i in weighted_list)
+		. += L
+	return
+
+///Return TRUE if you want thing to be attackable
+/datum/ai_controller/insane/murder/proc/CanTarget(atom/movable/thing)
+	var/mob/living/living_pawn = pawn
+	if(living_pawn.see_invisible < thing.invisibility)
+		return FALSE
+	if(isliving(thing))
+		var/mob/living/living_thing = thing
+		if(living_thing == living_pawn)
+			return FALSE
+		if(living_thing.status_flags & GODMODE)
+			return FALSE
+		if(living_thing.stat == DEAD)
+			return FALSE
+		if(!isturf(living_thing.loc) && !ismecha(living_thing.loc))
+			return FALSE
+		return TRUE
+	return FALSE
+
+///Checks for potential enemies in view with range, and sets the attack target for the ai; returns TRUE if target was set.
+/datum/ai_controller/insane/murder/proc/FindEnemies(range = aggro_range)
+	var/list/weighted_list = PossibleEnemies(range)
+	for(var/atom/movable/i in weighted_list)
+		//target type weight
 		if(istype(i, /mob/living/simple_animal/hostile))
-			weighted_list[i] = 3
+			weighted_list[i] = 4
 		else if(ishuman(i))
 			var/mob/living/carbon/human/H = i
 			if(H.sanity_lost)
-				weighted_list[i] = 2
-			else if(ismecha(H.loc))
 				weighted_list[i] = 3
+			else if(ismecha(H.loc))
+				weighted_list[i] = 4
 			else
-				weighted_list[i] = 5
+				weighted_list[i] = 7
 		else
 			weighted_list[i] = 1
+		//target distance weight
+		weighted_list[i] += 10 - min(get_dist(get_turf(pawn), get_turf(i)), 10)
+		//previous target weight
+		if(blackboard[BB_INSANE_CURRENT_ATTACK_TARGET] == i)
+			weighted_list[i] = max(weighted_list[i] - 10, 1)
 	if(weighted_list.len > 0)
-		blackboard[BB_INSANE_CURRENT_ATTACK_TARGET] = pickweight(weighted_list)
+		GiveTarget(pickweight(weighted_list))
 		return TRUE
 	return FALSE
 
+/datum/ai_controller/insane/murder/proc/GiveTarget(atom/movable/target)
+	blackboard[BB_INSANE_CURRENT_ATTACK_TARGET] = target
+	GainPatience()
+	current_movement_target = target
+	last_known_location = get_turf(target)
+	target_memory.Cut()
+	target_lost = FALSE
+
+/datum/ai_controller/insane/murder/proc/GainPatience()
+	if(lose_patience_timer_id)
+		deltimer(lose_patience_timer_id)
+		lose_patience_timer_id = null
+	lose_patience_timer_id = addtimer(CALLBACK(src, PROC_REF(TryChangeTarget)), lose_patience_time, TIMER_STOPPABLE)
+
+/datum/ai_controller/insane/murder/proc/TryChangeTarget()
+	lose_patience_timer_id = null
+	if(!FindEnemies())
+		for(var/datum/ai_behavior/insanity_attack_mob/i in current_behaviors)
+			i.finish_action(src, TRUE)
+
+/datum/ai_controller/insane/murder/proc/RegisterAggroValue(atom/movable/source, amount, damage_type)
+	if(QDELETED(source) || !damage_type)
+		return FALSE
+	if(!isnum(target_memory[source]))
+		target_memory += source
+
+/* 	if(damage_type == AGGRO_DAMAGE)
```

Comment:
Why is this commented out? I don't know. Could you add a comment explaining as to why?

Reply:
comment added